### PR TITLE
cherry pick the BYOID feature to v1.17

### PR DESCRIPTION
**Pull request by nan-yu**
This PR cherry picks the new `k8sserviceaccount` auth type feature into the v1.17 release branch, so that this feature can become available in 1.17.2.

This PR includes 4 dependent PRs: #1069, #1072, #1077 and #1078 to avoid cherry-pick conflicts caused by the fake client rotating service account key file.

This cherry-pick also requires manually resolving two conflicts:
1. When cherrypicking 6bb99a69472b7b63dd02cbb70ae260790d9b8167, `private_cert_secret_test.go` was modified by #1115: https://github.com/GoogleContainerTools/kpt-config-sync/pull/1115/files#diff-d11af36ecb44b307e149ad81de45e013e4362d32065532f3e716a708e596b879, #1104: https://github.com/GoogleContainerTools/kpt-config-sync/pull/1104/files#diff-d11af36ecb44b307e149ad81de45e013e4362d32065532f3e716a708e596b879, and #1111: https://github.com/GoogleContainerTools/kpt-config-sync/pull/1111/files#diff-d11af36ecb44b307e149ad81de45e013e4362d32065532f3e716a708e596b879.

   #1104 and #1111 include a new feature that is targeting the next minor release, so the conflicts were manually resolved.

2. When cherrypicking 1d50f52b6db62a1d699e197f0469d91e4f7d67d2, `rootsync_controller_test.go` and `reposync_controller_test.go` were modified by both #1121 and #1056. 

   #1056 added a `managed-by` label for manager, which can wait until the next patch release: https://github.com/GoogleContainerTools/kpt-config-sync/pull/1056/files#diff-4e452cd48e11abd5a08fa09505577300d00a536c7c57d7ebc492a9b166be3f55.

3. It also regenerates the deepcopy files by removing the `CACertSecretRef` field added in #1104 and #1111.

### CODE REVIEW

File: e2e/testcases/workload_identity_test.go [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+	if len(s) > l {
+		return s[:l]
+	}
+	return s
+}
+
+// migrateFromGSAtoKSA tests the scenario of migrating from impersonating a GSA
+// to leveraging KSA+WI (a.k.a, BYOID/Ubermint).
+func migrateFromGSAtoKSA(nt *nomostest.NT, rs *v1beta1.RootSync, ksaRef types.NamespacedName, fleetWITest bool, rootCommitFn nomostest.Sha1Func) error {
+	nt.T.Log("Update RootSync auth type from gcpserviceaccount to k8sserviceaccount")
```
[the remark this answers is not in the document]
Comment (nan-yu) [reply e2e/testcases/workload_identity_test.go:324]: Right. There's other missing e2e test coverage:
1. RepoSync with `gcpserviceaccount` for CSR
2. RepoSync with `gcenode`/`none` for CSR, OCI AR and Helm AR.
3. and more...

I can add a new PR to cover the `k8sserviceaccount` for RepoSync

**Comment by janetkuo** [comment]
`pkg/reconcilermanager/controllers/reposync_controller_test.go:3062:12: undefined: metadata.ConfigSyncManagedByLabel` this comes from the merge conflict

File: pkg/reconcilermanager/controllers/reposync_controller_test.go [hunk 2]
```diff
@@ -3040,18 +3040,38 @@ func TestRepoSyncWithHelm(t *testing.T) {
...
 		t.Fatalf("unexpected reconciliation error, got error: %q, want error: nil", err)
 	}
+
+	// test 1: validations
+	// No need to get RepoSync for a refreshed generation because the spec.override field is not updated.
+	labels := map[string]string{
+		metadata.SyncNamespaceLabel:       rs.Namespace,
+		metadata.SyncNameLabel:            rs.Name,
+		metadata.SyncKindLabel:            testReconciler.syncKind,
+		metadata.ConfigSyncManagedByLabel: reconcilermanager.ManagerName,
```
Comment (janetkuo) [inline pkg/reconcilermanager/controllers/reposync_controller_test.go:3062]: This is missed from the merge conflict resolution and causes presubmit failure `pkg/reconcilermanager/controllers/reposync_controller_test.go:3062:12: undefined: metadata.ConfigSyncManagedByLabel`

Comment (nan-yu) [reply pkg/reconcilermanager/controllers/reposync_controller_test.go]: After fixing this failure, there is another unit test error with unexpected RSync generation. That was caused by a missing dependent PR #1069. 

Basically, the change of `Update the fake clients to copy the new object state back into the input object.` in #1069 was available on the main branch, but not on the release branch, which caused the RSync generation not being updated with the PATCH call. After cherrypicking #1069, unit tests all passed.

**Review by tiffanny29631 — APPROVED** [review]
/lgtm

**Pull request merged**